addons/web/static/tests/core/overlay_service.test.js

```diff
@@ -24,6 +26,25 @@ test("simple case", async () => {
     expect(".o-overlay-container > .overlayed").toHaveCount(0);
 });
 
+test("shadow DOM overlays are visible when registered before main component is mounted", async () => {
+    class MyComp extends Component {
+        static template = xml`
+            <div class="overlayed"></div>
+        `;
+        static props = ["*"];
+    }
+    const env = await makeMockEnv();
+    const root = createElement("div");
+    getFixture().append(root);
+    root.setAttribute("id", "my-root-id");
+    getService("overlay").add(MyComp, undefined, { rootId: "my-root-id" });
+    const shadow = root.attachShadow({ mode: "open" });
+    const app = await mount(MainComponentsContainer, shadow, { env, getTemplate });
```

Comment:
`mountWithCleanup` doesn't support attaching to a specific target, which is necessary here so that this is attached on a shadow DOM.

We could improve `mountWithCleanup` function but since this PR is more a reactive response on a bug, I prefer to focus on the fix and have at least a good enough test coverage. Improving helpers is ok if trivially feasible and I know for sure the desired improvement, but that's not the case here so I prefer to just have a good enough solution, which consists of not using the helper and doing the mount + cleanup by hand.